source/common/router/scoped_rds.cc

```diff
@@ -121,19 +121,95 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::addOnDemandUpdateCallback(
+    std::function<void()> callback) {
+  // If route table has been initialized, run the callback to continue in filter chain, otherwise
+  // cache it and wait for the route table to be initialized.
+  if (route_provider_ != nullptr && !routeConfig()->name().empty()) {
+    callback();
+    return;
+  }
+  on_demand_update_callbacks_.push_back(callback);
+  // Initialize the rds provider if it has not been initialized.
+  maybeInitRdsConfigProvider();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::runOnDemandUpdateCallback() {
+  for (auto& callback : on_demand_update_callbacks_) {
+    callback();
+  }
+  on_demand_update_callbacks_.clear();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::initRdsConfigProvider(
+    envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
+    Init::Manager& init_manager) {
+  route_provider_ = std::dynamic_pointer_cast<RdsRouteConfigProviderImpl>(
+      parent_.route_config_provider_manager_.createRdsRouteConfigProvider(
+          rds, parent_.factory_context_, parent_.stat_prefix_, init_manager));
+
+  rds_update_callback_handle_ = route_provider_->subscription().addUpdateCallback([this]() {
+    // Subscribe to RDS update.
+    parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
+  });
+  parent_.stats_.active_scopes_.inc();
+}
 
-      rds_update_callback_handle_(route_provider_->subscription().addUpdateCallback([this]() {
-        // Subscribe to RDS update.
-        parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
-      })) {}
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::maybeInitRdsConfigProvider() {
+  // If the route provider have been initialized, return and wait for rds config update.
+  if (route_provider_ != nullptr) {
+    return;
+  }
+
+  // Create a init_manager to create a rds provider.
+  std::unique_ptr<Init::ManagerImpl> srds_init_mgr =
+      std::make_unique<Init::ManagerImpl>(fmt::format("SRDS on demand init manager."));
+  std::unique_ptr<Cleanup> srds_initialization_continuation =
+      std::make_unique<Cleanup>([this, &srds_init_mgr] {
+        Init::WatcherImpl noop_watcher(
+            fmt::format("SRDS on demand ConfigUpdate watcher: {}", scope_name_),
+            []() { /*Do nothing.*/ });
+        srds_init_mgr->initialize(noop_watcher);
+      });
+  // Create route provider.
+  envoy::extensions::filters::network::http_connection_manager::v3::Rds rds;
+  rds.mutable_config_source()->MergeFrom(parent_.rds_config_source_);
+  rds.set_route_config_name(
+      parent_.scoped_route_map_[scope_name_]->configProto().route_configuration_name());
+  initRdsConfigProvider(rds, *srds_init_mgr);
+  ENVOY_LOG(debug, fmt::format("Scope on demand update: {}", scope_name_));
+  // If RouteConfiguration hasn't been initialized, return.
```

Comment:
make it more clear here why we check on routeConfig.name()

Reply:
Sure, will add some comment.